pkg/ottl/ottlfuncs/func_replace_pattern.go

```diff
@@ -34,40 +35,86 @@ func createReplacePatternFunction[K any](_ ottl.FunctionContext, oArgs ottl.Argu
 		return nil, fmt.Errorf("ReplacePatternFactory args must be of type *ReplacePatternArguments[K]")
 	}
 
-	return replacePattern(args.Target, args.RegexPattern, args.Replacement, args.Function)
+	return replacePattern(args.Target, args.RegexPattern, args.Replacement, args.Function, args.IncludeNonCapture)
 }
 
-func applyOptReplaceFunction[K any](ctx context.Context, tCtx K, compiledPattern *regexp.Regexp, fn ottl.Optional[ottl.FunctionGetter[K]], originalValStr string, replacementVal string) (string, error) {
+func executeFunction[K any](ctx context.Context, tCtx K, compiledPattern *regexp.Regexp, fn ottl.Optional[ottl.FunctionGetter[K]], originalValStr string, replacementVal string, submatch []int) (string, error) {
+	result := compiledPattern.ExpandString([]byte{}, replacementVal, originalValStr, submatch)
+	fnVal := fn.Get()
+	replaceValGetter := ottl.StandardStringGetter[K]{
+		Getter: func(context.Context, K) (any, error) {
+			return string(result), nil
+		},
+	}
+	replacementExpr, errNew := fnVal.Get(&replacePatternFuncArgs[K]{Input: replaceValGetter})
+	if errNew != nil {
+		return "", errNew
+	}
+	replacementValRaw, errNew := replacementExpr.Eval(ctx, tCtx)
+	if errNew != nil {
+		return "", errNew
+	}
+	replacementValStr, ok := replacementValRaw.(string)
+	if !ok {
+		return "", fmt.Errorf("the replacement value must be a string")
+	}
+	return replacementValStr, nil
+}
+
+func applyOptReplaceFunction[K any](ctx context.Context, tCtx K, compiledPattern *regexp.Regexp, fn ottl.Optional[ottl.FunctionGetter[K]], originalValStr string, replacementVal string, includeNonCapture ottl.Optional[bool]) (string, error) {
 	var updatedString string
+	var captureGroups []string
+	var replacementValStr string
+	var err error
+	captureGroupMap := make(map[string]string)
 	updatedString = originalValStr
 	submatches := compiledPattern.FindAllStringSubmatchIndex(updatedString, -1)
+	// Extract capture group from replacement value to apply the function on it
+	r := regexp.MustCompile(`(\$[0-9])`)
```

Comment:
Do we need to support numbers with more than 1 digit?

Reply:
The thought here was that supporting $1 - $9 would be sufficient, since I haven't seen any regexes with that many capture groups.

Reply:
We should document we only support 9 max capture groups

Reply:
Added this to the README under ottlfuncs for `replace_pattern` and `replace_all_pattern`.